### ISSUE TITLE
False positive type-mismatch on braced match arm with write! tail expression

### ISSUE THREAD

**Issue by lonelam** [rust-lang/rust-analyzer#23039]
**rust-analyzer version**:  rust-analyzer version: 0.3.2997-standalone (b54a82b321 2026-08-02) 

**rustc version**: `rustc 1.91.1 (ed61e7d7e 2025-11-07)`

**editor or extension**: Cursor (VS Code-compatible), rust-analyzer extension 0.3.2997

**relevant settings**: defaults; no custom `rust-analyzer.*` settings required to reproduce

**repository link **: https://github.com/lonelam/rust-analyzer-weired-squiggle

<img width="722" height="360" alt="Image" src="https://github.com/user-attachments/assets/33be7f05-d3f8-4a07-bfbe-57193f8dd7a9" />

**code snippet to reproduce**:
```rust
use std::fmt;

pub enum E {
    A { reason: String },
    B { reason: String },
}

impl fmt::Display for E {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            // OK for rust-analyzer
            Self::A { reason } => write!(f, "{reason}"),
            // rust-analyzer false positive on the closing `}`:
            //   expected Result<(), Error>, found ()
            Self::B { reason } => {
                write!(f, "{reason}")
            }
        }
    }
}
```

## Summary

rust-analyzer's own type checker reports a type mismatch on a braced match arm whose tail expression is `write!(...)`, while `cargo check` accepts the code.

The diagnostic points at the arm's closing `}`:

```text
expected Result<(), Error>, found ()
```

Source is `rust-analyzer` (IDE analysis), not flycheck/`cargo check`.

## Expected

Both arms are equivalent and should type-check as `fmt::Result`. A block whose last expression is `write!(...)` (no semicolon) returns that `Result`.

## Actual

| Arm form | rustc | rust-analyzer |
| --- | --- | --- |
| `=> write!(f, "{reason}"),` | ok | ok |
| `=> { write!(f, "{reason}") }` | ok | false positive |

## Notes

- Zero dependencies; std only.
- Workaround: drop the braces so the arm is a single expression.
- Not the #3107 flycheck case — `cargo check` is clean; the squiggle is from RA's built-in diagnostics.
```

**Comment by ChayimFriedman2** [comment]
Your rustc version is not supported (you should've got a warning).
